File: src/main/java/selector/XPATHSelector.java
package selector;

import org.openqa.selenium.WebElement;
import org.openqa.selenium.chrome.ChromeDriver;
import org.openqa.selenium.firefox.FirefoxDriver;
import org.openqa.selenium.remote.RemoteWebDriver;

import static main.Starter.driver;
import static org.openqa.selenium.By.xpath;

public class XPATHSelector extends Selector {

    private static final XPATHSelector instance = new XPATHSelector();

    private XPATHSelector() {
    }

    public static XPATHSelector getInstance() {
        return instance;
    }

    public boolean elementExists(String selector) {
        return elementExists(driver, selector);
    }

    public boolean elementExists(RemoteWebDriver driver, String selector) {
        WebElement element = driver.findElement(xpath(selector));
        return super.elementExists(element);
    }

    public boolean click(String selector) {
        return click(driver, selector);
    }

    public boolean click(RemoteWebDriver driver, String selector) {
        WebElement element = driver.findElement(xpath(selector));
        return super.click(element);
    }


    public boolean setText(RemoteWebDriver driver, String selector, String text) {
        WebElement element = driver.findElement(xpath(selector));
        return super.setText(element, text);
    }

    public boolean setText(String selector, String text) {
        return setText(driver, selector, text);
    }

    public WebElement findElement(String selector) {
        return findElement(driver, selector);
    }

    public WebElement findElement(RemoteWebDriver driver, String selector) {
        WebElement element = driver.findElement(xpath(selector));
        return super.findElement(element);
    }

    //---------------------------------------------------------

    public RemoteWebDriver setFirefoxDriver(FirefoxDriver driver) {
        return super.setFirefoxDriver(driver);
    }

    public RemoteWebDriver setChromeDriver(ChromeDriver driver) {
        return super.setChromeDriver(driver);
    }

}

File: src/main/java/selector/XPATHSelector.kt
package selector

import org.openqa.selenium.By.xpath
import org.openqa.selenium.WebElement
import org.openqa.selenium.remote.RemoteWebDriver

class XPATHSelector(driver: RemoteWebDriver) : Selector(driver) {
    override fun findElement(selector: String): WebElement {
        return driver.findElement(xpath(selector))
    }

    override fun findAll(selector: String): List<WebElement> {
        val webElements = arrayListOf<WebElement>()
        var i = 1
        try {
            while (elementExists("""$selector[${i}]""")) {
                webElements.add(findElement("""$selector[${i}]"""))
                i++
                print("""Number of elements:""" + webElements.size + "\n")
            }
        } catch (e: Exception) {
            return webElements
        }
        return webElements
    }


    override fun elementExists(selector: String): Boolean {
        val element: WebElement = driver.findElement(xpath(selector))
        return super.elementExists(element)
    }

    override fun click(selector: String): Boolean {
        val element: WebElement = driver.findElement(xpath(selector))
        return super.click(element)
    }

    override fun setText(selector: String, text: String): Boolean {
        val element: WebElement = driver.findElement(xpath(selector))
        return super.setText(element, text)
    }

    override fun waitForElement(selector: String): Boolean {
        val element: WebElement = driver.findElement(xpath(selector))
        return super.waitForElement(element)
    }

}